File: BFS/Leetcode102.java

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if(root == null)
            return res;
        helper(root, res);
        return res;
    }
    
    private void helper(TreeNode root, List<List<Integer>> res) {
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while(!queue.isEmpty()) {
            List<Integer> levelList = new ArrayList<>();
            int size = queue.size();
            while(size > 0) {
                TreeNode temp = queue.poll();
                levelList.add(temp.val);
                if(temp.left != null)
                    queue.offer(temp.left);
                if(temp.right != null)
                    queue.offer(temp.right);
                size--;
                if(size == 0) {
                    res.add(levelList);
                    break;
                }  
            }
        }
    }
    
}

File: BFS/Leetcode102.js

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    const res = [], que = [];
    if(!root){
        return res;
    }
    que.push(root);
    let levelCount = 1;
    while(que.length) {
        levelCount = que.length;
        let levelArr = [];
        while(levelCount > 0) {
            const temp = que.shift();
            levelArr.push(temp.val);
            levelCount--;
            if(temp.left) {
                que.push(temp.left);
            }
            if(temp.right) {
                que.push(temp.right);
            }
        }    
        res.push(levelArr);
    }
    return res;
    
};